### ISSUE TITLE
Stopping a CohortExtractor instance results in duplicate messages

### ISSUE THREAD

**Issue by rkm** [SMI/SmiServices#829]
When CohortExtractor is shut down either by a signal (Ctrl+C) or through the control message queue, it's extremely likely that it will be midway through processing a single input message and will have sent on some amount of responses downstream. Since it exits immediately, the input message will be re-queued and the next service will end up sending duplicate messages.

While we can't be completely robust against message duplication (e.g. in the case of a complete crash), we can at least fix the service so that it finishes sending all the messages for a given input message before exiting. The machinery for this should be similar to other services e.g. DicomTagReader.

**Comment by tznind** [comment]
Wouldn't it be easier to make the downstream services deal with duplication correctly? e.g. by overwritting a file in the output directory?

**Comment by jas88** [comment]
Yes, idempotence makes this a non-issue; a "clean" shutdown is usually neater when possible, so having the ctrl-C/SIGTERM handler set a flag and quit abruptly on _second_ invocation seems best for that? Retrying some messages could result in a lot of extra work (like redoing a whole 1TB extraction that had been 99% complete): we need to make sure doing that doesn't actually _break_ anything, but should try to minimise how often it happens too.
